modules/currency-l0/src/main/scala/org/tessellation/currency/l0/snapshot/services/StateChannelBinarySender.scala

```diff
@@ -0,0 +1,195 @@
+
+  def make[F[_]: Async: Hasher: Temporal](
+    pendingR: Ref[F, List[TrackedBinary]],
+    expectedOrdinalsToConfirmR: Ref[F, PosLong],
+    avgConfirmedCountPerSnapshotR: Ref[F, NonNegLong],
+    catchUpMode: Ref[F, Boolean],
+    identifierStorage: IdentifierStorage[F],
+    globalL0ClusterStorage: L0ClusterStorage[F],
+    lastGlobalSnapshotStorage: LastSnapshotStorage[F, GlobalIncrementalSnapshot, GlobalSnapshotInfo],
+    stateChannelSnapshotClient: StateChannelSnapshotClient[F]
+  ): StateChannelBinarySender[F] =
+    new StateChannelBinarySender[F] {
+
+      private val logger = Slf4jLogger.getLogger
+
+      def process(binary: Hashed[StateChannelSnapshotBinary]): F[Unit] =
+        catchUpMode.get.ifM(
+          enqueue(binary),
+          enqueue(binary) >> post(binary)
+        )
+
+      def confirm(globalSnapshot: GlobalIncrementalSnapshot): F[Unit] = for {
+        identifier <- identifierStorage.get
+        currentOrdinal <- lastGlobalSnapshotStorage.getOrdinal
+        binaries = globalSnapshot.stateChannelSnapshots.get(identifier).toList.flatMap(_.toList)
+        confirmedHashesInGlobalSnapshot <- binaries.traverse(_.toHashed).map(_.flatMap(b => List(b.hash, b.lastSnapshotHash)))
+
+        (confirmed, pending) <- pendingR.modify { prevPending =>
+          val (confirmedWithIndex, unconfirmedWithIndex) = prevPending.zipWithIndex.partition {
+            case (TrackedBinary(tracked, _), _) => confirmedHashesInGlobalSnapshot.contains(tracked.hash)
+          }
+
+          val unconfirmed = unconfirmedWithIndex.map { case (binary, _) => binary }
+          val confirmed = confirmedWithIndex.map { case (binary, _) => binary }
+          val pending = confirmedWithIndex.lastOption.map { case (_, index) => index }
+            .map(cuttingIndex => unconfirmedWithIndex.collect { case (binary, index) if index >= cuttingIndex => binary })
+            .getOrElse(unconfirmed)
+
+          (pending, (confirmed, pending))
+        }
+
+        expectedOrdinalsToConfirm <- updateAvgExpectedOrdinalsToConfirm(confirmed, globalSnapshot.ordinal)
+        _ <- updateAvgConfirmedCountPerSnapshot(confirmed)
+        _ <- updateCatchUpMode(pending, currentOrdinal, expectedOrdinalsToConfirm)
+
+        _ <- logger
+          .warn(
+            s"${pending.length} state channel binaries are still not confirmed on the global state. Waiting for the next global snapshot."
+          )
+          .whenA(pending.nonEmpty)
+      } yield ()
+
+      def processPending: F[Unit] = {
+        val retry = for {
+          pending <- pendingR.get
+          avgConfirmedPerSnapshot <- avgConfirmedCountPerSnapshotR.get
+          cap = Math
+            .round(avgConfirmedPerSnapshot.toDouble + (avgConfirmedPerSnapshot.toDouble / 2))
+            .toInt // Try sending 50% more than avg confirmed
+          toRetry = pending.take(cap)
+          _ <- toRetry.traverse(tracked => enqueue(tracked.binary))
```

Comment:
Seems like you are enqueuing the binaries once more rather than resending them. The only moment you are using `post` function is in `process` and only if we are not in catchup mode.

Reply:
Good catch. Previously both methods were merged so I somehow missed it.